### PR TITLE
Track current debouncing state

### LINKED ISSUE #40: Weird Edge-Casey Render Race
I feel like I'm probably running into some issues that are specific to my particular use case.  I ended up forking and fixing them [here](https://github.com/nkbt/react-debounce-input/compare/master...ericf89:master), but it probably is too specific to my case to be merged in.  Thought I'd let you look it over though just in case. 😄 

I have a situation where my debounced input is re-rendered (by a component higher up the tree) while it's debouncing its onChange handler.   This means it receives props, sets state, and renders with it's `props` value which is normally a couple keystrokes behind the user, before triggering it's change handler that properly updates the value of the input in the store, which trickles all the way back down again. 

The result is a kinda flicker back and forth between the correct value and the one in props as everything stabilizes.  

I 'fixed' it by keeping track of whether or not the debounced input is currently `debouncing`. If so, it ignores `value` changes in its props, (under the assumption that it'll change again when the debounced handler actually fires/notifies.  I don't really like ignoring the props change...  but it seemed like the simplest way to fix my issue. 💀 

Let me know if you think this is actually useful, and I can formalize it into a PR, or you can cannibalize it how you see fit. 

If not you can close this, and I'll use my version for my use case.   Cheers!

### PR DESCRIPTION
**Pull request by ericf89**
I believe this fixes #40 as least as far as my particular use case is concerned. 

Let me know if you have suggestions about the `createNotifier` func... It's a little meatier now that I'm trying to keep track of the debouncing state, but hopefully it still makes sense.  Other suggestions or requests are of course welcome too! 

Thanks! 🎉

### CODE REVIEW

File: src/Component.js [hunk 1]
```diff
@@ -59,31 +62,45 @@ export const DebounceInput = React.createClass({
 
 
   componentWillUnmount() {
-    if (this.notify.cancel) {
-      this.notify.cancel();
+    if (this.notify.flush) {
+      this.notify.flush();
```
Comment (nkbt) [inline src/Component.js]: need to reset  `isDebouncing` on flush too

File: src/Component.js [hunk 2]
```diff
@@ -46,6 +46,9 @@ export const DebounceInput = React.createClass({
 
 
   componentWillReceiveProps({value, debounceTimeout}) {
+    if (this.isDebouncing) {
```
Comment (nkbt) [inline src/Component.js]: I will need to check if behaviour is not changed critically with this change. There are several conrollable inputs around (react-filter-input and some in projects). That might take some time.

**Comment by nkbt** [comment]
Thanks a lot for the PR, looks pretty good for me, I need to check some edge cases tho...

File: src/Component.js (hunk 2 above)
Comment (ericf89) [inline src/Component.js]: Yea, this is the scariest part.... Totally understand. No rush.

If you want, I can hide this behind an optional prop to make this a little safer backwards-compatibility-wise.

File: src/Component.js (hunk 1 above)
Comment (ericf89) [inline src/Component.js]: I figured it probably wasn't too critical when the component is on it's way out, but I'll add it anyway for consistency.  👍

File: src/Component.js (hunk 2 above)
Comment (nkbt) [inline src/Component.js]: Nah, let it be the default behaviour. If it breaks anything, I'll just publish a new major version instad.

**Comment by scandinaro** [comment]
@nkbt I'm running into this issue as well.  Any chance this is ready to merge?

**Comment by nkbt** [comment]
May it be resolved with #59 by any chance (published ad react-debounce-input@2.4.1)?

**Comment by scandinaro** [comment]
@nkbt I just upgraded from 2.4.0 to 2.4.1 and still have the same issue

**Comment by nkbt** [comment]
Ok, will rebase this pr, test and merge

**Comment by scandinaro** [comment]
@nkbt can this be merged yet?

**Comment by nkbt** [comment]
Haven't had a chance to get here, travelling now... this needs to be rebased with conflicts, so cannot be merged from mobile yet

**Comment by ericf89** [comment]
I'll try to rebase this soon and see how it all looks.

**Comment by nkbt** [comment]
Thanks!

**Comment by ericf89** [comment]
@nkbt @scandinaro   I rebased this to the best of my ability. 😅   

My brain is a little out of context as I have't followed along with the recent changes, so definitely will want a bit of regression testing.   Let me know if anything seems weird.

**Comment by nkbt** [comment]
Thanks heaps for your work, @ericf89!
I did some checks on couple of other projects and everything seems to work well as far as I can tell!

**Comment by nkbt** [comment]
Published react-debounce-input@2.4.2

**Comment by huan086** [comment]
An alternative implementation would be to save `event.target.value` as `oldValue` at `doNotify`, and check if `oldValue === nextProps.value` at `componentWillReceiveProps`. Logic being, if I notified parent of this value, and I get back this value as nextProps, I can safely ignore it. Thus, only 2 places needs to be changed. Did I miss out any cases for this?

Advantage of the alternative approach is that when some other events causes parent to set another value of debounce-input while debouncing, the input will reflect the value (instead of ignoring it).

File: src/Component.js (hunk 2 above)
Comment (nkbt) [inline src/Component.js]: Oops it actually broke `forceUpdate` with `infinite` timeout (-1)

**Pull request merged**